### PR TITLE
Add time to HGCAL CaloParticles

### PR DESCRIPTION
**Pull request by AuroraPerego**
#### PR description:

Following the `MtdTruthAccumulator.cc` logic, the `SimVertex` time has also been added to the HGCAL `CaloParticle`s. 
This is needed because the information coming from the `SimVertexCollection` of the `CaloParticle`s from pileup is lost if not saved in the `CaloTruthAccumulator.cc`.
The name of the method used to assign the time has been changed to the more appropriate `setSimTime`.
The correction of the time for oot pileup has been removed since the `SimVertex` time takes that already into account.

#### PR validation:

Tested on wfs 29888.203 (classical mixing) and 29888.99 (premix) enabling the `CaloParticle`s from pileup (they are disabled by default). 
Example to test it:
```
test parameters:
    - workflow_opts= -w upgrade
    - workflow = 29888.99,29888.203
```
FYI @waredjeb @felicepantaleo @rovere

### CODE REVIEW

**Comment by fwyzard** [comment]
allow @AuroraPerego test rights

File: SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc [hunk 1]
```diff
@@ -223,6 +229,7 @@ CaloTruthAccumulator::CaloTruthAccumulator(const edm::ParameterSet &config,
     : messageCategory_("CaloTruthAccumulator"),
       maximumPreviousBunchCrossing_(config.getParameter<unsigned int>("maximumPreviousBunchCrossing")),
       maximumSubsequentBunchCrossing_(config.getParameter<unsigned int>("maximumSubsequentBunchCrossing")),
+      bunchSpacing_(config.getParameter<unsigned int>("bunchspace")),
```
Comment (Moanwar) [inline SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc:232]: Hi, just a small comment—I don't see the usage of bunchSpacing. Maybe I'm missing something? Thanks

**Comment by AuroraPerego** [comment]
test parameters:
    - workflow_opts= -w upgrade
    - workflow = 29888.203

**Comment by AuroraPerego** [comment]
please test

File: RecoHGCal/TICL/plugins/SimTrackstersProducer.cc [hunk 2]
```diff
@@ -293,7 +287,7 @@ void SimTrackstersProducer::produce(edm::Event& evt, const edm::EventSetup& es)
     // Create a Trackster from the object entering HGCal
     if (cp.g4Tracks()[0].crossedBoundary()) {
       regr_energy = cp.g4Tracks()[0].getMomentumAtBoundary().energy();
-      float time = cp.g4Tracks()[0].getPositionAtBoundary().t();
+      float time = cp.g4Tracks()[0].getPositionAtBoundary().t() * 1e9;
```
[the remark this answers is not in the document]
Comment (AuroraPerego) [reply RecoHGCal/TICL/plugins/SimTrackstersProducer.cc]: It's from seconds to ns, I've changed it to use the constant `s` from CLHEP.

File: SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc (hunk 1 above)
Comment (AuroraPerego) [reply SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc]: Yes sorry, it was a leftover from previous changes. I've removed it.

File: RecoHGCal/TICL/plugins/SimTrackstersProducer.cc [hunk 3]
```diff
@@ -32,15 +32,15 @@
...
-#include "SimDataFormats/Vertex/interface/SimVertex.h"
-
 #include "SimDataFormats/TrackingAnalysis/interface/TrackingParticle.h"
 #include "SimDataFormats/TrackingAnalysis/interface/UniqueSimTrackId.h"
 
 #include "SimDataFormats/Associations/interface/TrackToTrackingParticleAssociator.h"
 
 #include "DataFormats/HGCalReco/interface/Common.h"
 
+#include "CLHEP/Units/GlobalSystemOfUnits.h"
```
Comment (fwyzard) [inline RecoHGCal/TICL/plugins/SimTrackstersProducer.cc:42]:
```suggestion
#include <CLHEP/Units/GlobalSystemOfUnits.h>
```

File: RecoHGCal/TICL/plugins/SimTrackstersProducer.cc [hunk 4]
```diff
@@ -293,7 +289,8 @@ void SimTrackstersProducer::produce(edm::Event& evt, const edm::EventSetup& es)
     // Create a Trackster from the object entering HGCal
     if (cp.g4Tracks()[0].crossedBoundary()) {
       regr_energy = cp.g4Tracks()[0].getMomentumAtBoundary().energy();
-      float time = cp.g4Tracks()[0].getPositionAtBoundary().t();
+      // G4 time is in seconds (s), convert to ns
+      float time = cp.g4Tracks()[0].getPositionAtBoundary().t() * s;
```
[the remark this answers is not in the document]
Comment (fwyzard) [reply RecoHGCal/TICL/plugins/SimTrackstersProducer.cc:293]: also, this seems quite counter-intuitive: you start from `t()` that is expressed in seconds and multiply by "seconds" to get the time in nanoseconds ?

**Comment by AuroraPerego** [comment]
please test

**Comment by AuroraPerego** [comment]
please test

**Comment by felicepantaleo** [comment]
ignore tests-rejected with ib-failure

**Comment by civanch** [comment]
+1

**Comment by jfernan2** [comment]
+1

**Comment by Moanwar** [comment]
+Upgrade

**Comment by mandrenguyen** [comment]
+1

**Comment by rovere** [comment]
This PR changes the class layout of both `CaloParticle` and `MTDCaloParticle` w/o increasing the class version in the proper xml file.
Aren't we catching this kind of compilation warnings?

**Comment by AuroraPerego** [comment]
> Aren't we catching this kind of compilation warnings?

Apparently, we are not now (`scram b` gives no warnings/errors). I ran `scram b updateclassversion` and it did not fail, while `edmCheckClassVersion` generated the new checksum correctly.
I'll test the phase2 HLT timing test with the new checksum, but there is probably an issue with `scram` not catching it (?).
Maybe @smuzaffar has more insights into this.

**Comment by mmusich** [comment]
>  I'll test the phase2 HLT timing test with the new checksum, but there is probably an issue with scram not catching it (?).

I tested it fixes the phase2 HLT timing and opened https://github.com/cms-sw/cmssw/pull/46851

**Comment by smuzaffar** [comment]
https://github.com/cms-sw/cmsdist/pull/9543 should properly catch mismatching class versions

**Pull request merged**